src/keyword_vpd_parser.cpp

```diff
@@ -0,0 +1,143 @@
+#include "keyword_vpd_parser.hpp"
+
+#include "constants.hpp"
+#include "exceptions.hpp"
+
+#include <iostream>
+#include <numeric>
+#include <string>
+
+namespace vpd
+{
+
+/* @brief Keyword VPD parser function*/
+std::variant<types::ParsedVPD, types::KeywordVpdMap> KeywordVpdParser::parse()
+{
+    try
+    {
+    m_vpdIterator = m_keywordVpdVector.begin();
+
+    if(*m_vpdIterator != constants::KW_VPD_START_TAG)
+    {
+        throw(DataException("Invalid Large resource type Identifier String"));
+    }
+
+    checkNextBytesValidity(sizeof(constants::KW_VPD_START_TAG));
+    std::advance(m_vpdIterator, sizeof(constants::KW_VPD_START_TAG));
+
+    uint16_t dataSize = getKwDataSize();
+
+    checkNextBytesValidity(constants::TWO_BYTES + dataSize);
+    std::advance(m_vpdIterator, constants::TWO_BYTES + dataSize);
+
+    // Check for invalid vendor defined large resource type
+    if (*m_vpdIterator != constants::KW_VPD_START_TAG)
+    {
+        if (*m_vpdIterator != constants::ALT_KW_VPD_START_TAG)
+        {
+            throw(DataException("Invalid Keyword Vpd Start Tag"));
+        }
+    }
+    auto kwValMap = parser();
+
+    // Do these validations before returning parsed data.
+    // Check for small resource type end tag
+    if (*m_vpdIterator != constants::KW_VAL_PAIR_END_TAG)
+    {
+        throw(DataException("Invalid Small resource type End"));
+    }
+
+    validateChecksum();
+
+    // Check vpd end Tag.
+    if (*m_vpdIterator != constants::KW_VPD_END_TAG)
+    {
+        throw(DataException("Invalid Small resource type."));
+    }
+
+    return kwValMap;
+    }
+    catch (const std::exception& e)
+    {
+        //if (typeid(e) == std::type_index(typeid(DataException)))
+        {
+	}
+
+	return types::KeywordVpdMap{};
+    }
+}
+
+types::KeywordVpdMap KeywordVpdParser::parser()
+{
+    m_checkSumStart = m_vpdIterator;
+    checkNextBytesValidity(constants::ONE_BYTE);
+    std::advance(m_vpdIterator, constants::ONE_BYTE);
+
+    auto totalSize = getKwDataSize();
+    if (totalSize == 0)
+    {
+        throw(DataException("Data size is 0, badly formed keyword VPD"));
+    }
+    
+    types::KeywordVpdMap kwValMap;
+
+    checkNextBytesValidity(constants::TWO_BYTES);
+    std::advance(m_vpdIterator, constants::TWO_BYTES);
+
+    // Parse the keyword-value and store the pairs in map
+    while (totalSize > 0)
+    {
+        checkNextBytesValidity(constants::TWO_BYTES);
+        std::string keywordName(m_vpdIterator,
```

Comment:
Move it next to the place where it is being used.

Reply:
yes

Reply:
Okay.